src/AbstractApiClient.php

```diff
@@ -0,0 +1,266 @@
+     * @throws WrongResponseException Response in unexpected format or structure
+     */
+    public function put(string $uri, array $body = [], array $query = [], array $headers = [])
+    {
+        return $this->send((new Request('PUT', $uri)), $body, $query, $headers);
+    }
+
+    /**
+     * @param string $uri
+     * @param array $body
+     * @param array $query
+     * @param array $headers
+     * @return array
+     * @throws BadResponseException For 5xx and unhandled 4xx
+     * @throws ErrorResponseException For handled 4xx
+     * @throws TransportException For errors on transport layer
+     * @throws WrongResponseException Response in unexpected format or structure
+     */
+    public function patch(string $uri, array $body = [], array $query = [], array $headers = [])
+    {
+        return $this->send((new Request('PATCH', $uri)), $body, $query, $headers);
+    }
+
+    /**
+     * @param string $uri
+     * @param array $query
+     * @param array $headers
+     * @return array
+     * @throws BadResponseException For 5xx and unhandled 4xx
+     * @throws ErrorResponseException For handled 4xx
+     * @throws TransportException For errors on transport layer
+     * @throws WrongResponseException Response in unexpected format or structure
+     */
+    public function delete(string $uri, array $query = [], array $headers = [])
+    {
+        return $this->send((new Request('DELETE', $uri)), [], $query, $headers);
+    }
+
+    public function getClient(): ClientInterface
+    {
+        return $this->client;
+    }
+
+    protected function setClient(ClientInterface $client): self
+    {
+        $this->client = $client;
+
+        return $this;
+    }
+
+    protected function getLogger(): ?LoggerInterface
+    {
+        return $this->logger;
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param array $body
+     * @param array $query
+     * @param array $headers
+     * @return array
+     * @throws BadResponseException
+     * @throws TransportException
+     * @throws WrongResponseException
+     * @throws ErrorResponseException
+     */
+    private function send(RequestInterface $request, array $body = [], array $query = [], array $headers = [])
+    {
+        $options = $this->processRequestBody($body);
+        $this->processQuery($query, $options);
+        $this->processHeaders($headers, $options);
+
+        try {
+            $response = $this->getClient()->send($request, $options);
+
+            return $this->processResponse($response, $request);
+        } catch (GuzzleException $e) {
+            $this->handleException($e, $request);
+        }
+    }
+
+    /**
+     * @param GuzzleException $e
+     * @param RequestInterface $request
+     * @throws BadResponseException
+     * @throws ErrorResponseException
+     * @throws TransportException
+     */
+    protected function handleException(GuzzleException $e, RequestInterface $request)
```

Comment:
Making a note of http://docs.guzzlephp.org/en/stable/quickstart.html#exceptions in a comment may be helpful, but I'm not that fussed about it. (The content may change as they update, and can cause comment sync issues, so meh, actually forget about it)